pkg/proxy/iptables/proxier_test.go

```diff
@@ -6833,114 +6833,574 @@ func TestEndpointSliceWithTerminatingEndpointsTrafficPolicyCluster(t *testing.T)
-		COMMIT
-		`)
-	assertIPTablesRulesEqual(t, getLine(), expected, fp.iptablesData.String())
+		fp.syncProxyRules()
+	}
 
-	runPacketFlowTests(t, getLine(), fp.iptablesData.String(), testNodeIP, []packetFlowTest{
+	// We use the same flowTests for all of the testCases. The "output" and "masq"
+	// values here represent the normal case (working localDetector, no masqueradeAll)
+	flowTests := []packetFlowTest{
 		{
-			name:     "pod to cluster IP",
+			name:     "pod to ClusterIP",
 			sourceIP: "10.0.0.2",
 			destIP:   "172.30.0.41",
 			destPort: 80,
-			output:   "10.180.0.1:80",
+			output:   "10.180.0.1:80, 10.180.1.1:80",
+			masq:     false,
+		},
+		{
+			name:     "pod to NodePort",
+			sourceIP: "10.0.0.2",
+			destIP:   testNodeIP,
+			destPort: 3001,
+			output:   "10.180.0.1:80, 10.180.1.1:80",
 			masq:     true,
 		},
 		{
 			name:     "pod to LB",
 			sourceIP: "10.0.0.2",
 			destIP:   "1.2.3.4",
 			destPort: 80,
-			output:   "10.180.0.1:80",
+			output:   "10.180.0.1:80, 10.180.1.1:80",
+			masq:     true,
+		},
+		{
+			name:     "node to ClusterIP",
+			sourceIP: testNodeIP,
+			destIP:   "172.30.0.41",
+			destPort: 80,
+			output:   "10.180.0.1:80, 10.180.1.1:80",
+			masq:     true,
+		},
+		{
+			name:     "node to NodePort",
+			sourceIP: testNodeIP,
+			destIP:   testNodeIP,
+			destPort: 3001,
+			output:   "10.180.0.1:80, 10.180.1.1:80",
+			masq:     true,
+		},
+		{
+			name:     "localhost to NodePort",
+			sourceIP: "127.0.0.1",
+			destIP:   "127.0.0.1",
+			destPort: 3001,
+			output:   "10.180.0.1:80, 10.180.1.1:80",
 			masq:     true,
 		},
 		{
-			name:     "external to cluster IP",
+			name:     "node to LB",
+			sourceIP: testNodeIP,
+			destIP:   "1.2.3.4",
+			destPort: 80,
+			output:   "10.180.0.1:80, 10.180.1.1:80",
+			masq:     true,
+		},
+		{
+			name:     "external to ClusterIP",
 			sourceIP: testExternalClient,
 			destIP:   "172.30.0.41",
 			destPort: 80,
-			output:   "10.180.0.1:80",
+			output:   "10.180.0.1:80, 10.180.1.1:80",
+			masq:     true,
+		},
+		{
+			name:     "external to NodePort",
+			sourceIP: testExternalClient,
+			destIP:   testNodeIP,
+			destPort: 3001,
+			output:   "10.180.0.1:80, 10.180.1.1:80",
 			masq:     true,
 		},
 		{
 			name:     "external to LB",
 			sourceIP: testExternalClient,
 			destIP:   "1.2.3.4",
 			destPort: 80,
-			output:   "10.180.0.1:80",
+			output:   "10.180.0.1:80, 10.180.1.1:80",
 			masq:     true,
 		},
-	})
+		{
+			name:     "pod to ClusterIP with eTP:Local",
+			sourceIP: "10.0.0.2",
+			destIP:   "172.30.0.42",
+			destPort: 80,
+
+			// externalTrafficPolicy does not apply to ClusterIP traffic, so same
+			// as "Pod to ClusterIP"
+			output: "10.180.0.2:80, 10.180.1.2:80",
+			masq:   false,
+		},
+		{
+			name:     "pod to NodePort with eTP:Local",
+			sourceIP: "10.0.0.2",
+			destIP:   testNodeIP,
+			destPort: 3002,
+
+			// FIXME: The short-circuit rule means we potentially send to a remote
```

Comment:
I'm not sure I agree, but I am not sure I don't.

When `eTP:Cluster` we always masquerade (at least we always have in the past) , I think that's generally agreed, right?  I don't see why we would change that.

We decided local source + `eTP:Local` is a special case.  If we want to emulate up-and-out-and-back-down, masquerade would be right for external LBs (e.g. things that leave a cloud's VPC) but would be wrong for internal LBs (e.g. in the local network).

I *think* I'd rather err on the side of "errantly retain information" than "errantly lose information".

Is there some case where the masquerade in this case is semantically important?

Reply:
I wasn't saying "we should masquerade in the eTP:Local case", I was saying "either the masquerading in the Cluster case is necessary (in which case Local is broken), OR the masquerading in the Cluster case is not necessary (in which case why are we doing it?)"

I was assuming we wanted to avoid masquerading any time we could avoid it, but I guess "we always masquerade all traffic to EXT in the Cluster case, so therefore we masquerade pod-to-EXT traffic" is a plausible argument...

Reply:
> I was assuming we wanted to avoid masquerading any time we could avoid it, but I guess "we always masquerade all traffic to EXT in the Cluster case, so therefore we masquerade pod-to-EXT traffic" is a plausible argument...

Yeah, I have always assumed that it would more confusing to avoid SNAT in some cases but not others.  Ergo "When eTP:Cluster we always masquerade" even if it isn't strictly necessary for local clients.

Plausible?